azure-client-authentication/src/main/java/com/microsoft/azure/credentials/ApplicationTokenCredentials.java

```diff
@@ -132,17 +140,54 @@ String clientCertificatePassword() {
         return clientCertificatePassword;
     }
 
+    /**
+     * Gets the timeout for AAD authentication calls. Default is 60 seconds.
+     *
+     * @return the timeout in seconds.
+     */
+    public long timeoutInSeconds() {
+        return timeoutInSeconds;
+    }
+
+    /**
+     * Sets the timeout for AAD authentication calls. Default is 60 seconds.
+     *
+     * @param timeoutInSeconds the timeout in seconds.
+     * @return the modified ApplicationTokenCredentials instance
+     */
+    public ApplicationTokenCredentials withTimeoutInSeconds(long timeoutInSeconds) {
+        this.timeoutInSeconds = timeoutInSeconds;
+        return this;
+    }
+
     @Override
-    public synchronized String getToken(String resource) throws IOException {
+    public String getToken(String resource) throws IOException {
         AuthenticationResult authenticationResult = tokens.get(resource);
         if (authenticationResult == null || authenticationResult.getExpiresOnDate().before(new Date())) {
-            authenticationResult = acquireAccessToken(resource);
+            ReentrantLock lock;
+            synchronized (authenticationLocks) {
+                lock = authenticationLocks.get(resource);
+                if (lock == null) {
+                    lock = new ReentrantLock();
+                    authenticationLocks.put(resource, lock);
+                }
+            }
+            lock.lock();
+            try {
+                authenticationResult = tokens.get(resource);
+                if (authenticationResult == null || authenticationResult.getExpiresOnDate().before(new Date())) {
+                    tokens.put(resource, acquireAccessToken(resource).get(timeoutInSeconds, TimeUnit.SECONDS));
+                }
+            } catch (Exception e) {
+                throw new IOException(e.getMessage(), e);
+            } finally {
+                lock.unlock();
+            }
         }
-        tokens.put(resource, authenticationResult);
         return authenticationResult.getAccessToken();
     }
 
-    private AuthenticationResult acquireAccessToken(String resource) throws IOException {
+    private Future<AuthenticationResult> acquireAccessToken(String resource) throws IOException {
         String authorityUrl = this.environment().activeDirectoryEndpoint() + this.domain();
         ExecutorService executor = Executors.newSingleThreadExecutor();
```

Comment:
@jianghaolu now that this function can return before `context.acquireToken` finishes, and it shuts down the executor before returning (L221), can we guarantee that `context.acquireToken` does not submit new tasks to the executor asynchronously? In that case, the async tasks might be rejected when `acquireAccessToken` returns early.

It seems that the interface doesn't guarantee this cannot happen, and the correctness of the code depends on the implementation detail of `AuthenticationContext.acquireToken`. Can we at least add a test to make sure we don't accidentally break the assumption in the future?

Reply:
Thanks for catching this! It does seem like the executor can be shut down before the Future gets executed. Let me fix this and add a unit test.
I tested the current code with AAD but a smoke test is probably insufficient for real-world workloads.